레스토랑 상세 페이지 구현하기

2scent opened the pull request:
안녕하세요. 이번 한 주도 잘 부탁드립니다 😄

현재 `HomePage`, `AboutPage`, `RestaurantsPage`, `NotFoundPage`로 분리하고, 라우팅하는 것까지 진행했습니다.

이번 과제부터 페이지도 생기다 보니 파일이 많아져서 디렉터리별로 나눠서 관리할 필요성이 느껴지는데요.

예전에 리액트를 공부할 때도 꽤 고민했던 내용인데 디렉터리 구조를 짜는 건 항상 어렵더라고요. 트레이너님의 노하우나 관련 자료 공유해주실 수 있을까요?

2scent committed:
Add react-router-dom

2scent committed:
Edit App for rendering Home

기존의 App 기능들은 RestaurantsPage로 분리
App에서는 HomePage를 보여주도록 수정

2scent committed:
Edit App for using react-router-dom

2scent committed:
Add about page

2scent committed:
Edit webpack dev server settings

존재하지 않는 페이지에 접근하면 index.html로 redirect 되도록 수정

2scent committed:
Add App test for routing about page

2scent committed:
Add header

2scent committed:
Edit App for routing the restaurants page

2scent committed:
Add not found page

2scent committed:
Fix typo in App test

with invalid paths일 때, not found page를 렌더하도록 테스트명 수정

daadaadaah commented on src/App.jsx:
```diff
@@ -1,32 +1,26 @@
+      </h1>
+      <Routes>
+        <Route index element={<HomePage />} />
+        <Route path="about" element={<AboutPage />} />
+        <Route path="restaurants" element={<RestaurantsPage />} />
```
path를 `/about`, `/restaurants` 이렇게도 쓸 수 있어요!

2scent replied:
트레이너님의 피드백을 보고 `/`의 역할에 대해 공부해봤습니다.

[https://reactrouter.com/docs/en/v6/getting-started/concepts#definitions](https://reactrouter.com/docs/en/v6/getting-started/concepts#definitions)

> Relative links - Links that don't start with `/` will inherit the closest route in which they are rendered. This makes it easy to link to deeper URLs without having to know and build up the entire path.

요약하면, `/`를 붙이면 절대 경로가 되고, 붙이지 않으면 상대 경로가 된다는 건데요. 지금처럼 최상위 경로일 때는 딱히 차이가 없지만, 하위 경로일 때는 차이가 생길 수 있습니다.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="/detail" element={<DetailPage />} />
</Route>
```

이렇게 작성하면, `/detail`은 절대 경로가 되어 

`Uncaught Error: Absolute route path "/detail" nested under path "/about" is not valid. An absolute child route path must start with the combined path of all its parent routes.`

와 같이 절대 경로인 `/detail`은 `/about`의 하위 경로가 될 수 없다는 에러가 발생하는데요.

이를 해결하기 위해서는 두 가지 방법이 있습니다.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="/about/detail" element={<Detail />} />
</Route>
```

이 방법은 하위 경로에서 `/about`을 중복적으로 작성해야 하기 때문에 좋지 않은 것 같고요.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="detail" element={<Detail />} />
</Route>
```

이렇게 하위 경로에서 `/`를 붙이지 않는 방법이 있습니다.

결론을 말하자면, 하위 경로에서 `/`를 붙이지 않을 거면 전체적으로 `/`를 붙이지 않음으로써 일관성을 유지하는 게 낫다고 생각합니다.

2scent replied:
트레이너님의 피드백을 보고 `/`의 역할에 대해 공부해봤습니다.

[https://reactrouter.com/docs/en/v6/getting-started/concepts#definitions](https://reactrouter.com/docs/en/v6/getting-started/concepts#definitions)

> Relative links - Links that don't start with `/` will inherit the closest route in which they are rendered. This makes it easy to link to deeper URLs without having to know and build up the entire path.

요약하면, `/`를 붙이면 절대 경로가 되고, 붙이지 않으면 상대 경로가 된다는 건데요. 지금처럼 최상위 경로일 때는 딱히 차이가 없지만, 하위 경로일 때는 차이가 생길 수 있습니다.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="/detail" element={<Detail />} />
</Route>
```

이렇게 작성하면, `/detail`은 절대 경로가 되어 

`Uncaught Error: Absolute route path "/detail" nested under path "/about" is not valid. An absolute child route path must start with the combined path of all its parent routes.`

와 같이 절대 경로인 `/detail`은 `/about`의 하위 경로가 될 수 없다는 에러가 발생하는데요.

이를 해결하기 위해서는 두 가지 방법이 있습니다.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="/about/detail" element={<Detail />} />
</Route>
```

이 방법은 하위 경로에서 `/about`을 중복적으로 작성해야 하기 때문에 좋지 않은 것 같고요.

```javascript
<Route path="/about" element={<AboutPage />}>
  <Route path="detail" element={<Detail />} />
</Route>
```

또 다른 방법으로 하위 경로에서 `/`를 붙이지 않는 방법이 있습니다.

결론을 말씀드리면, 하위 경로에서 `/`를 붙이지 않을 거면 전체적으로 `/`를 붙이지 않음으로써 일관성을 유지하는 게 낫다고 생각해서 현재 코드를 유지하기로 했습니다.

daadaadaah replied:
피드백에 대해 한번 생각해보고 공부해보시는거 너무 좋은 프로세스인 것 같아요!👍
결국, 중첩 라우팅까지 고려해봤을 때, 일관성있게 상대경로로 관리하는게 더 나을 것 같다는 의견으로 이해하면 될까용?ㅎㅎ
좋습니다!👍 저는 계속 절대경로로만 관리해서, 상대경로로 했을 때 어떻게 될지 궁금하네요!ㅎㅎ

2scent replied:
제가 글을 좀 급하게 마무리했었네요. 이해하신 게 맞습니다 👍

daadaadaah reviewed (changes requested):
잘하셨어요!👍 계속 진행해주세요~ 🚀

> 예전에 리액트를 공부할 때도 꽤 고민했던 내용인데 디렉터리 구조를 짜는 건 항상 어렵더라고요. 트레이너님의 노하우나 관련 자료 공유해주실 수 있을까요?

저도 프로젝트 할 때마다 디렉터리 구조에 대해 고민을 많이 하는 편인데, 그냥 팀이 편한 구조 또는 내가 편한 구조가 좋은 구조인 것 같아요! 예를 들어, 팀 단위가 도메인별로 나눠져 있으면, 도메인별로 폴더구조를 가져가는게 좋고, 그렇지 않고 많은 팀원들이 기능별로 나눠서 짜는게 익숙하면, 기능별로 나눠서 짜는게 좋은 것 같아요!ㅎㅎ
react 공식문서에서도 특별히 추천하는 디렉터리 구조는 없지만 몇가지 인기 있는 구조들은 있다고 하네요! 아래 링크 한번 참고해보세요!
참고 : https://ko.reactjs.org/docs/faq-structure.html

2scent committed:
Add restaurant page

라우팅도 추가

2scent committed:
Add Restaurants

RestaurantsContainer에서 뷰 로직을 Restaurants로 분리

2scent committed:
Edit directory structure

- pages 폴더 만들고, 연관된 컴포넌트끼리 분류
  - RestaurantPage -> RestaurantDetailPage
- redux 관련 파일들은 store 폴더로

2scent committed:
Add fetchRestaurant

2scent committed:
Add setRestaurant action

2scent committed:
Separate async actions

2scent committed:
Add loadRestaurant action

2scent committed:
Add RestaurantContainer

2scent committed:
Add Restaurant

2scent committed:
Add msw with settings

2scent committed:
Fix restaurant loading error

RestaurantContainer에서 restaurant를 fetch 중이라 비어 있을 때 로딩
표시를 보여주도록 수정

2scent commented:
디렉토리 구조에 대해 알려주신 자료 포함해서 추가적인 학습을 하고 나름대로 적용해봤는데 아직 익숙하지 않아서 그런지 정제되지 못한 느낌이 있네요. 남은 기간 동안 계속해서 리팩토링을 해보겠습니다.

과제의 기본적인 요구 사항은 얼추 완성한 것 같은데 로딩과 관련된 로직은 리덕스에서 처리할 수 있도록 리팩토링 할 생각입니다.

디스코드로 알려주신 Mock Service Worker 이용해서 정상 동작하는지까지 확인했습니다.

daadaadaah commented on src/App.test.jsx:
```diff
@@ -21,14 +21,72 @@ test('App', () => {
     restaurants: [
       { id: 1, name: '마법사주방' },
     ],
+    restaurant: {
+      id: 1,
+      name: '김밥제국',
+      category: '분식',
+      address: '서울시 강남구 역삼동',
+      menus: [
+        { id: 1, name: '김밥' },
+        { id: 2, name: '라면' },
+      ],
+    },
```
fixtures에 있는 데이터를 활용해도 좋을 것 같아요!

daadaadaah commented on src/pages/restaurant-detail/Restaurant/Restaurant.test.jsx:
```diff
@@ -0,0 +1,33 @@
+    />
+  ));
+
+  it('renders the restaurant name', () => {
+    const { container } = renderRestaurant();
+
+    expect(container).toHaveTextContent(RESTAURANT.name);
+  });
+
+  it('renders the restaurant address', () => {
+    const { container } = renderRestaurant();
+
+    expect(container).toHaveTextContent(RESTAURANT.address);
+  });
+
+  it('renders the restaurant menus', () => {
+    const { container } = renderRestaurant();
+
+    RESTAURANT.menus.forEach(
+      ({ name }) => expect(container).toHaveTextContent(name),
+    );
+  });
```
"render restaurant info" 이런 식으로 하나의 테스트로 묶어도 괜찮을 거 같아요!

daadaadaah commented on src/pages/restaurant-detail/Restaurant/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    restaurant: RESTAURANT,
+  }));
+
+  it('renders fetched restaurant', () => {
```
`fetched` 단어 빼도 테스트 이해하는데 문제 없을 것 같아요!

daadaadaah commented on src/pages/restaurants/RestaurantsPage.jsx:
```diff
@@ -0,0 +1,32 @@
+  loadInitialData,
+} from '../../store/async-actions';
+
+// 0. 지역, 분류 목록을 얻기
+// 1. 지역 선택 - Regions <- API (0)
+// 2. 분류 선택 - Categories - 한식, 중식, 일식, ... <- API (0)
+// 3. 식당 목록 - Restaurants <- API (with region, category) -> 1, 2 모두 완료된 경우
```
불필요한 주석 삭제해줍시당!ㅎㅎ

2scent committed:
Use fixtures for tests

fixtures 폴더에 index.js를 추가해서 모듈화 하고,
여러 테스트에서 fixtures를 사용하도록 함

2scent committed:
Edit Restaurant test

불필요하게 나눠져 있는 테스트를 하나로 통합

2scent committed:
Edit test name

2scent committed:
Remove unnecessary comments

2scent committed:
Add isLoading state to store

2scent committed:
Use setLoading for async actions

2scent committed:
Refactor async actions

비동기 작업 전후로 setLoading 액션을 반복적으로 호출하는 것을 함수로
묶음

2scent committed:
Fix wrong initial state for reducer test

2scent committed:
Edit RestaurantContainer for using isLoading

2scent committed:
Add absolute path settings

2scent committed:
Use absolute import

2scent commented:
진행 내용

- 피드백 반영
- 로딩 관련 로직은 리덕스에서 처리
- import 할 때 뎁스가 깊어지는 것을 회피하기 위해 절대 경로를 사용

---

리덕스에서 `isLoading` 상태를 관리하고, 비동기 액션 전후에 `isLoading`을 `true` 및 `false`로 바뀌게 하고, `RestaurantContainer`에서 로딩 화면을 보여줄지 정하기 위해 `isLoading`을 이용하도록 했습니다.

현재 문제가 `loadRestaurant` 액션이 실행돼야 `isLoading`이 `true`로 바뀌는데, 컴포넌트가 마운트 된 후에 `loadRestaurant` 액션이 실행되다 보니 마운트 직후에는 `isLoading`이 `false` 값을 가져서, 로딩 화면을 표시할 때 여전히 다음과 같이 `restaurant`가 비었는지 여부를 확인해야 하는데요. 뭔가 잘못된 것 같은데 어떻게 개선할 수 있을까요?

```javascript
if (isLoading || !restaurant) return <p>로딩 중</p>;
```

---

> 컴포넌트에서 다른 파일 import할 때, 뎁스가 깊어지는 불편함이 있네요.

뎁스가 깊어지는 것을 회피하기 위해 절대 경로를 세팅했는데 eslint 설정값도 수정해서 CI 테스트를 통과할 수 있을지 모르겠네요. 현재 각 모듈들의 구성(e.g. index.js 활용 등)이나 네이밍(파스칼 케이스 or 케밥 케이스)에 대해 고민 중인데 다음 PR까지 정리해보겠습니다.

daadaadaah commented:
1. isLoading 
- loadRestaurant 을 page 컴포넌트에서 실행시켜 보세요!

2. 폴더구조
- 오! 절대 경로를 사용해서 문제를 해결하셨군요~잘하셨어요👍
- eslint 설정값 수정으로 `lint기 통과 될까?`가 고민이시면, 로컬에서 `npm run lint` 해보시면 될 것 같아요! 

3. 네이밍
- 네이밍 같은거 오래 고민하지 마시구~ 그냥 일반적으로 React 진영에서 많이 쓰이는 네이밍(컴포넌트 파일 같은 경우는 파스칼 케이스, 변수명은 카멜 케이스 등등)으로 하시는 걸 추천드려요! 어짜피 네이밍은 팀에서 팀원이랑 같이 정하는 것이므로, 네이밍에 고민하는데 많은 시간 안쓰셨으면 좋겠습니다!

daadaadaah commented on package.json:
```diff
@@ -40,10 +41,15 @@
     "webpack-dev-server": "^4.8.1"
   },
   "dependencies": {
+    "msw": "0.44.0",
```
msw가 언제 필요하지에 대해 생각해보면, `devDependencies`에서 관리하는게 좀더 적절할 것 같은데, 어떻게 생각하시나요?ㅎㅎ

2scent replied:
원래 `devDependencies`로 추가했다가 eslint에서 막길래 그냥 `dependencies`로 추가했었는데 eslint 설정 수정하고, 다시 `devDependencies`로 변경했습니다.

2scent committed:
Use msw as devDependencies

2scent committed:
Edit not found message for e2e test

api도 실제 서버에 요청하도록 수정

2scent committed:
Edit Restaurant for real data

2scent committed:
Edit main.js path to absolute path

2scent committed:
Edit initial isLoading state to be true

2scent committed:
Edit async actions

loadRestaurant만 isLoading을 변경하도록 수정

2scent committed:
Refactor RestaurantContainer

isLoading만 확인해서 로딩 중 렌더하도록 수정

2scent committed:
Update folder namgin in pages

각 페이지 폴더들 이름을 파스칼 케이스로 변경
폴더마다 index.js를 추가해서 폴더명으로 import 하도록 함

2scent commented:
> loadRestaurant 을 page 컴포넌트에서 실행시켜 보세요!

page에서 처리하려고 해봤는데 같은 문제가 발생해서, `isLoading`의 초기값을 `true`로 지정하고, `loadRestaurant`에서만 수정되게 했습니다. 각 액션에 따라 로딩 상태를 가지려면 상태별로 `store`를 나눠야 할 때가 되지 않았나 싶네요. 다만 지금도 에러가 발생하진 않지만, A 레스토랑을 조회한 후에 B 레스토랑을 조회한다면 그 전의 A 레스토랑 정보가 깜빡이면서 잠깐 보이는 문제가 있는데 이건 어떻게 해결할 수 있을지 잘 모르겠네요.

네이밍은 말씀하신 것처럼 파스칼 케이스로 했고, 폴더마다 `index.js`를 둬서 보다 깔끔하게 import 되도록 했습니다. 이와 관련해서 한 가지 질문이 있는데요.

`pages/RestaurantDetailPage/Restaurant/index.js`를 보면 default로 `RestaurantContainer`만 export 했는데요. 그래서 `RestaurantDetailPage`에서는 `import Restaurant from './Restaurant';`와 같은 형식으로 해당 컴포넌트가 Container인지 아닌지 모르고 그냥 쓸 수 있도록 했는데 괜찮은 방법일까요?

daadaadaah commented:
> page에서 처리하려고 해봤는데 같은 문제가 발생해서, `isLoading`의 초기값을 `true`로 지정하고, `loadRestaurant`에서만 수정되게 했습니다. 각 액션에 따라 로딩 상태를 가지려면 상태별로 `store`를 나눠야 할 때가 되지 않았나 싶네요. 다만 지금도 에러가 발생하진 않지만, A 레스토랑을 조회한 후에 B 레스토랑을 조회한다면 그 전의 A 레스토랑 정보가 깜빡이면서 잠깐 보이는 문제가 있는데 이건 어떻게 해결할 수 있을지 잘 모르겠네요.

- 어제 문제를 해결하고 코멘트를 수정한다는 게 깜박했네요ㅠㅠ🙏 `page` 컴포넌트가 아닌 아래코드처럼  `RestaurantsContainer`에서 loadRestaurant를 dispatch 하고, setRestaurant와 setLoading 의 dispatch 순서를 바꿔보세요! 이렇게 로직이 바뀌게 되면, 지금 깜박인 문제도 해결할 수 있을 거예요!
<img width="667" alt="스크린샷 2022-07-16 오후 11 57 49" src="https://user-images.githubusercontent.com/60481383/179360222-2018c154-c183-4cb6-8532-5a6df2faa29d.png">
<img width="667" alt="스크린샷 2022-07-17 오전 12 05 28" src="https://user-images.githubusercontent.com/60481383/179360551-556716a3-6c20-4b4c-a56a-443bc46e689a.png">


> 네이밍은 말씀하신 것처럼 파스칼 케이스로 했고, 폴더마다 `index.js`를 둬서 보다 깔끔하게 import 되도록 했습니다. 이와 관련해서 한 가지 질문이 있는데요.
> 
> `pages/RestaurantDetailPage/Restaurant/index.js`를 보면 default로 `RestaurantContainer`만 export 했는데요. 그래서 `RestaurantDetailPage`에서는 `import Restaurant from './Restaurant';`와 같은 형식으로 해당 컴포넌트가 Container인지 아닌지 모르고 그냥 쓸 수 있도록 했는데 괜찮은 방법일까요?

- 음...제가 1번문제 직접 해결해보려고 2scent 코드 pull 받아서 작업해본 결과, Container라고 명시 안해줘서 헷갈렸어요ㅠㅠ 각 컴포넌트를 사용하는 쪽에서 이 컴포넌트가 정확히 무슨 컴포넌트인지는 명시하는 것이 좋을 것 같아요!

2scent commented:
> 어제 문제를 해결하고 코멘트를 수정한다는 게 깜박했네요ㅠㅠ🙏 `page` 컴포넌트가 아닌 아래코드처럼  `RestaurantsContainer`에서 loadRestaurant를 dispatch 하고, setRestaurant와 setLoading 의 dispatch 순서를 바꿔보세요! 이렇게 로직이 바뀌게 되면, 지금 깜박인 문제도 해결할 수 있을 거예요!

`RestaurantsContainer`에서만 loadRestaurant를 dispatch 하게 되면, `/restaurants/1`에 바로 접근했을 때 문제가 생길 텐데요. 이를 해결하기 위해 `RestaurantContainer`에서는 현재 `restaurant`가 비어있을 때만 dispatch 하도록 하면 될까요?

daadaadaah commented:
> `RestaurantsContainer`에서만 loadRestaurant를 dispatch 하게 되면, `/restaurants/1`에 바로 접근했을 때 문제가 생길 텐데요. 

`RestaurantsContainer`에서도 loadRestaurant를 dispatch해주면 된다는 의미였어요!ㅎㅎ
그런데, 맨처음 난 에러가 아래처럼 reducer에서 restaurant의 속성값 셋팅을 안해줘서 난 에러였네요!
이렇게 셋팅해주면, 굳이 `RestaurantsContainer`에서도 loadRestaurant를 dispatch해줘도 될 것 같아요!
<img width="667" alt="스크린샷 2022-07-17 오전 12 58 47" src="https://user-images.githubusercontent.com/60481383/179362475-b7ea47ed-73a4-4a53-8cff-c30d104c2287.png">

그리고,
> A 레스토랑을 조회한 후에 B 레스토랑을 조회한다면 그 전의 A 레스토랑 정보가 깜빡이면서 잠깐 보이는 문제
저는 발생하지 않는데, 지금도 발생하는지 확인부탁드려요!

2scent committed:
Edit initial restaurant to be non-null

2scent committed:
Rename component folders to have container

2scent committed:
Edit the restaurant detail e2e test

2scent commented:
오늘은 다음 두 가지 작업 진행했습니다.

- `restaurant`에 초기값 할당
- 컴포넌트 폴더에 Container 붙이기

> A 레스토랑을 조회한 후에 B 레스토랑을 조회한다면 그 전의 A 레스토랑 정보가 깜빡이면서 잠깐 보이는 문제

이 문제는 일단 그대로 나뒀는데요. [이 피드백](https://github.com/CodeSoom/react-week6-assignment-1/pull/97#issuecomment-1186216618)에서 주신 방법대로 하면 해결은 되나 다음과 같이 두 가지 문제점이 있을 수 있다고 생각합니다.

- 의도하신 바가 `<Link onClick={handleClick}>` 같은 형태라 생각되는데요. Link 컴포넌트의 본 역할이 아닌 데이터를 가져오는 기능이 추가돼서 혼란이 생길 수 있을 것 같습니다.
- 어플리케이션이 커져서 `레스토랑 상세 페이지`와 연결되는 페이지가 여러 개가 된다면, 그 페이지마다 `loadRestaurant`를 호출해야 할 텐데 그러면 디버깅을 할 때 `loadRestaurant`이 어디서 호출되는지 추적이 어려울 수 있을 것 같습니다.

이런 이유로 `loadRestaurant`는 그 데이터를 사용하는 `RestaurantContainer` 혹은 그 페이지에서 호출하고, 깜빡이는 문제는 로딩 상태에 대한 접근 방법 자체를 바꿀 필요가 있다는 생각이 드는데 트레이너님은 어떻게 생각하시나요?

daadaadaah commented:
> `loadRestaurant`는 그 데이터를 사용하는 `RestaurantContainer` 혹은 그 페이지에서 호출하고, 깜빡이는 문제는 로딩 상태에 대한 접근 방법 자체를 바꿀 필요가 있다는 생각이 드는데 트레이너님은 어떻게 생각하시나요?

저는 좋다고 생각해요! 다른 요소를 변경함으로써 문제를 해결하는 모습 너무 좋네요👍 

여담으로, 제가 직접 짠 코드는 아니지만 저도 다른 개발자분들과 만나서 개발 이야기 하다가 들은 건대, 
어떤 걸 클릭해서 상세보기 페이지로 넘어가는 것을 구현할 때, 클릭하는 요소에서 상세보기 페이지에서 필요한 데이터를 불러오고 난 후 화면이 넘어가도록 구현하는 경우가 있더라구요! 그냥 이렇게 구현하는 방법도 있다 정도만 알고 계시면 좋을 것 같아요! 기회가 되면 나중에 한번 활용해보시면 좋구요!

추가로, 이번주 TIL에 보면, 코테 준비중이시던대, 코드숨으로 TDD 익숙해지셨으면, 코테할 때도 한번 적용시켜보시는건 어떠신가요? 저 같은 경우에는 코테 할 때 TDD로 하니, Green 떴을 때 성취감도 있고, 무엇을 해야하는지 명확해서 저는 좋더라구요! 

이번주 한주도 고생 많으셨어요~ 2scent 님 덕분에 이번주 저도 많이 배웠네요! 감사합니당!🙏

주간회고를 통해 한주를 한번 되돌아봅시당!🤔
